spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/query/NearQuery.java

```diff
@@ -420,25 +426,54 @@ public DBObject toDBObject() {
 		}
 
 		if (maxDistance != null) {
-			dbObject.put("maxDistance", maxDistance.getNormalizedValue());
+			dbObject.put("maxDistance", getDistanceValueInRadiantsOrMeters(maxDistance));
 		}
 
 		if (minDistance != null) {
-			dbObject.put("minDistance", minDistance.getNormalizedValue());
+			dbObject.put("minDistance", getDistanceValueInRadiantsOrMeters(minDistance));
 		}
 
 		if (metric != null) {
-			dbObject.put("distanceMultiplier", metric.getMultiplier());
+			dbObject.put(
+					"distanceMultiplier",
+					usesMetricSystem() ? BigDecimal.ONE.divide(getMetricSystemNormalizer(metric), PRECISION + 3,
+							RoundingMode.HALF_UP).divide(new BigDecimal(1000), PRECISION + 3,
+							RoundingMode.HALF_UP).doubleValue() : metric.getMultiplier());
 		}
 
 		if (num != null) {
 			dbObject.put("num", num);
 		}
 
-		dbObject.put("near", Arrays.asList(point.getX(), point.getY()));
+		if (point instanceof GeoJsonPoint) {
```

Comment:
That `instance of` check is used twice here and encapsulated in a method below. I think we should reuse the method to express we're referring to the same conceptual thing. I am not sure I understand enough about the issue being solved here but the need to introduce a lot of private methods makes me feel like we want to introduce a value object to express the underlying concepts more explicitly. Those cascades of `BigDecimal` operations feel rather low level compared to what we're trying to achieve. Could this be operations on `Distance` maybe?